src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -783,6 +795,17 @@ private void TrackNullableStateForAssignment(BoundExpression value, TypeSymbolWi
                         }
                     }
                 }
+                else if (targetType.IsNullableType())
+                {
+                    Debug.Assert(targetType.TypeSymbol.Equals(valueType.TypeSymbol, TypeCompareKind.AllIgnoreOptions));
```

Comment:
>Debug.Assert(targetType.TypeSymbol.Equals(valueType.TypeSymbol, TypeCompareKind.AllIgnoreOptions)); [](start = 20, length = 99)

I do not think the assert makes this code robust enough. For cases above (reference types) and below (regular structs), we actually make sure the types match before we attempt to copy state. I think we should do an explicit check here as well.